simphony_mayavi/sources/slim_cuds_source.py

```diff
@@ -0,0 +1,232 @@
+    """Small generator that processes an iterable of strings
+    into an iterable of CUBA keys"""
+    for elem in iterable:
+        try:
+            yield CUBA[elem]
+        except ValueError:
+            pass
+
+
+def _available_keys(cuds):
+    """Given a cuds, it returns a tuple of sets, containing the available
+    CUBA keys divided in classes:
+
+        - point_scalars
+        - point_vectors
+        - cell_scalars
+        - cell_vectors
+
+    Parameters
+    ----------
+    cuds :
+        A cuds data source
+
+    Returns
+    -------
+    _ : tuple(set, set, set, set)
+        A tuple of four sets for each of the classes above.
+    """
+
+    point_scalars = set()
+    point_vectors = set()
+    cell_scalars = set()
+    cell_vectors = set()
+
+    if isinstance(cuds, (ABCMesh, H5Mesh)):
+        for point in cuds.iter_points():
+            scalar_keys, vector_keys = _extract_cuba_keys_per_data_types(
+                point.data)
+            point_scalars.update(scalar_keys)
+            point_vectors.update(vector_keys)
+
+        for element in itertools.chain(cuds.iter_cells(),
+                                       cuds.iter_edges(),
+                                       cuds.iter_faces()):
+            scalar_keys, vector_keys = _extract_cuba_keys_per_data_types(
+                element.data)
+            cell_scalars.update(scalar_keys)
+            cell_vectors.update(vector_keys)
+
+    elif isinstance(cuds, ABCParticles):
+        for particle in cuds.iter_particles():
+            scalar_keys, vector_keys = _extract_cuba_keys_per_data_types(
+                particle.data)
+            point_scalars.update(scalar_keys)
+            point_vectors.update(vector_keys)
+
+        for bond in cuds.iter_bonds():
+            scalar_keys, vector_keys = _extract_cuba_keys_per_data_types(
+                bond.data)
+            cell_scalars.update(scalar_keys)
+            cell_vectors.update(vector_keys)
+
+    elif isinstance(cuds, ABCLattice):
+        for node in cuds.iter_nodes():
+            scalar_keys, vector_keys = _extract_cuba_keys_per_data_types(
+                node.data)
+            point_scalars.update(scalar_keys)
+            point_vectors.update(vector_keys)
+
+    else:
+        msg = 'Provided object {} is not of any known cuds type'
+        raise TraitError(msg.format(type(cuds)))
+
+    return point_scalars, point_vectors, cell_scalars, cell_vectors
+
+
+def _extract_cuba_keys_per_data_types(data):
+    """Given a DataContainer, gets the scalar and vector CUBA types
+    it contains, as two independent sets
+
+    Parameters
+    ----------
+    data : DataContainer
+        A data container.
+
+    Returns
+    -------
+    scalars, vectors : tuple(set, set)
+        The CUBA type keys for scalar and vector data, respectively
+    """
+    scalars = set()
+    vectors = set()
+
+    for cuba_key in data.keys():
+        try:
+            dim = _cuba_dimensionality(cuba_key)
```

Comment:
You can probably get the dimensionality of the CUBA key from `KEYWORDS`, https://github.com/simphony/simphony-common/blob/master/simphony/core/keywords.py